contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,119 @@
+/// RandomBeaconHistory (FLIP 123)
+///
+/// This contract stores the history of random sources generated by the Flow network. The defined Heartbeat resource is
+/// updated by the Flow Service Account at the end of every block with that block's source of randomness.
+///
+/// While the source values are random, generated by the Random Beacon and transmitted into the execution 
+/// environment via the committing transaction, using the raw values from this contract does not guarantee secure
+/// randomness. The Hearbeat is intended to be used in conjunction with a pseudo-random number generator (PRNG)
+/// construct and commit-reveal mechanism to provide an onchain source of secure, non-revertible randomness.
+///
+/// Read the full FLIP here: https://github.com/onflow/flips/pull/123
+///
+pub contract RandomBeaconHistory {
+
+    /* --- Canonical Paths --- */
+    //
+    pub let HeartbeatStoragePath: StoragePath
+    pub let HeartbeatPublicPath: PublicPath
+
+    /* --- Events --- */
+    //
+    pub event HearbeatInitialized(atBlockHeight: UInt64)
+    pub event NewSourceOfRandomness(atBlockHeight: UInt64, randomSource: [UInt8])
+
+    /* --- Hearbeat --- */
+    //
+    /// A publicly queryable interface for the Heartbeat resource.
+    ///
+    pub resource interface HeartbeatPublic {
+        pub fun getRandomSourceHistory(): [[UInt8]]
+        pub fun getInitHeight(): UInt64
+        pub fun sourceOfRandomness(atBlockHeight blockHeight: UInt64): [UInt8]
+    }
+
+    /// The Heartbeat resource containing each block's source of randomness in sequence
+    ///
+    pub resource Heartbeat : HeartbeatPublic {
+        /// The height at which the first source of randomness was recorded
+        access(self) var initHeight: UInt64?
+        /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+        // TODO: Test at scale
+        access(self) let randomSourceHistory: [[UInt8]]
```

Comment:
this should be implemented as a limited length cyclic array.

- There should be a max length for it
- When writing a new value you overwrite the oldest value, and you keep the index of the latest block stored somewhere
- We might also need admin functions (on a admin resource) to increase and decrease the max length of the array.

Reply:
My understanding from @dete's comments [here](https://github.com/onflow/flips/pull/123#issuecomment-1693863438) was that we would store `randomSourceHistory` indefinitely. If that's not the path forward, I can proceed with initial plan to implement a cyclic array.

Reply:
Ah, of course! forgot about that.

Reply:
I think reading single value from this will cost like ~100 computation usage after a year.

Reply:
Yes the FLIP got updated to propose a non-limited history.